File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
// recursive dfs - null node has height of -1
// dfs returns height, used to calc diameter
const diameterOfBinaryTree = function(root) {
    let maxDiameter = 0;
    
    const dfs = node => {
        if (!node) return -1;
        
        const left = dfs(node.left),
              right = dfs(node.right),
              diameter = 2 + left + right;
        
        maxDiameter = Math.max(maxDiameter, diameter);
        
        return 1 + Math.max(left, right);
    }
    
    dfs(root);
    return maxDiameter;
};

File: easy/543-diameter-of-binary-tree/543-diameter-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
    res = 0
    
    dfs = -> (node) do 
        return 0 if !node
        
        l = dfs.call node.left
        r = dfs.call node.right
        res = [res, l + r].max
        
        1 + [l, r].max
    end
    
    dfs.call root
    
    res
end